File: grind75/week1/055-Validate-Binary-Search-Tree-Medium/055-Validate-Binary-Search-Tree-Medium.py
# Question link: https://leetcode.com/problems/validate-binary-search-tree/description/
# Think of this problem in terms of the lower and upper bound that a node must be within


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def validate(node, left, right):
            if node is None:
                return True
            
            if node.val <= left or node.val >= right:
                return False

            return (
                validate(node.left, left, node.val) and
                validate(node.right, node.val, right)
            )
        

        return validate(root, float('-inf'), float('inf'))

        

File: grind75/week1/055-Validate-Binary-Search-Tree-Medium/055-Validate-Binary-Search-Tree-Medium.cpp
// Problem link: https://leetcode.com/problems/validate-binary-search-tree/description/

// Problem link: https://leetcode.com/problems/validate-binary-search-tree/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    bool valid(TreeNode* node, long minimum, long maximum)
    {
        if(node == NULL) { return true; }

        if(node->val >= maximum || node->val <= minimum)
        {
            return false;
        }

        return valid(node->left, minimum, node->val) && valid(node->right, node->val, maximum);
    }

    bool isValidBST(TreeNode* root) {
        return valid(root, numeric_limits<long>::min(), numeric_limits<long>::max());
    }
};
